### ISSUE TITLE
ci: K8sUpdates failing for Kernel 4.9 during downgrade operations

### ISSUE THREAD

**Issue by nbusseneau** [cilium/cilium#15059]
Upgrade/downgrade CI testing from K8s 1.14 up to 1.18 on Kernel 4.9 has been failing on `master` for quite some time:
```
20:19:39  • Failure [405.272 seconds]
20:19:39  K8sUpdates
20:19:39  /home/jenkins/workspace/cilium-master-k8s-1.18-kernel-4.9/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:478
20:19:39    Tests upgrade and downgrade from a Cilium stable image to master [It]
20:19:39    /home/jenkins/workspace/cilium-master-k8s-1.18-kernel-4.9/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:514
20:19:39  
20:19:39    migrate-svc restart count values do not match
[2021-02-20T19:19:39.466Z]   Expected
[2021-02-20T19:19:39.466Z]       <int>: 0
[2021-02-20T19:19:39.466Z]   to be identical to
[2021-02-20T19:19:39.466Z]       <int>: 7
20:19:39  
20:19:39    /home/jenkins/workspace/cilium-master-k8s-1.18-kernel-4.9/src/github.com/cilium/cilium/test/k8sT/Updates.go:379
```
(example from https://jenkins.cilium.io/job/cilium-master-k8s-1.18-kernel-4.9/808/, and [here's the ZIP](https://github.com/cilium/cilium/files/6023701/ff0e9e8f_K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master.zip))

**Comment by Skymirrh** [comment]
I have reproduced the issue locally. The failing check is on line 557 here, after downgrading version:
https://github.com/cilium/cilium/blob/51924951ecafb26dcdbeb1f9a39a0f1522986936/test/k8sT/Updates.go#L537-L561

This checks the restart count of pods to sniff flow interruption:
https://github.com/cilium/cilium/blob/51924951ecafb26dcdbeb1f9a39a0f1522986936/test/k8sT/Updates.go#L353-L381

And indeed pods are restarting while downgrading.

**Comment by Skymirrh** [comment]
So, a bit more of investigation yields the following:
- This specific test was passing until Feb 8: https://jenkins.cilium.io/view/Obsolete/job/cilium-master-K8s-all/589/
- Started to fail on every run starting Feb 9: https://jenkins.cilium.io/view/Obsolete/job/cilium-master-K8s-all/590/
- Didn't get caught as it was in the middle of a series of failures (Feb 8 and previous: flakes, Feb 10 and next: Jenkinsfiles split was implemented and brought other issues to the light), and didn't get the amount of attention it should have.
- Retrieved list of PRs merged on Feb 8: https://github.com/cilium/cilium/pulls?q=merged%3A2021-02-08..2021-02-09
- Filtered obvious non-culprits and ordered potential culprits, with #14853 as highest candidate.
- Bingo! Apparently the changes to preallocate map flags introduced the issue (checked by testing on commit right before PR is merged and right after, twice).
- Didn't get caught by CI because required jobs were all passing. The only failing jobs are the optional K8s versions (1.14 - 1.18) on Kernel 4.9, Kernel 4.19 and net-next are not affected, neither is 1.20 on Kernel 4.9.

So now the question is how do we fix that? I am a complete newbie with BPF.

**Comment by christarazi** [comment]
Two things here.

1) K8s version should not matter; only the kernel version. https://github.com/cilium/cilium/pull/14853 affects maps that used specific map types that were not supported in 4.9 and thus the wrong map type was used to determine the flags, causing them to be recreated.

You mentioned that K8s 1.20 & kernel 4.9 was a combo that was _not_ failing?

2) I _think_ this is expected, I'm trying to convince myself that this is the case. The above PR fixes the upgrade path. If a user downgrades, they're going back to the buggy behavior that determines the flags, based on the incorrectly detected map type. 

I've marked the aforementioned PR with `upgrade-impact` label so that we can properly track this when it comes time to include it in a release. See https://github.com/cilium/cilium/pull/14853#issuecomment-784602294.

**Comment by Skymirrh** [comment]
> You mentioned that K8s 1.20 & kernel 4.9 was a combo that was not failing?

Indeed, as can be seen either from [automated jobs](https://jenkins.cilium.io/job/cilium-master-k8s-1.20-kernel-4.9/) or [PR jobs](https://jenkins.cilium.io/view/PR/job/Cilium-PR-K8s-1.20-kernel-4.9/) all passing without issue. I've confirmed it on local runs as well.

> I think this is expected, I'm trying to convince myself that this is the case. The above PR fixes the upgrade path. If a user downgrades, they're going back to the buggy behavior that determines the flags based on the incorrectly detected map type.

That makes sense to me, but in this case are we supposed to keep the test failing as-is?

**Comment by christarazi** [comment]
> Indeed, as can be seen either from automated jobs or PR jobs all passing without issue. I've confirmed it on local runs as well.

That's very strange. :thinking: 

>That makes sense to me, but in this case are we supposed to keep the test failing as-is?

**Comment by pchaigno** [comment]
> You mentioned that K8s 1.20 & kernel 4.9 was a combo that was not failing?

It is not failing because the K8sUpdates test doesn't run on K8s-1.20-Kernel-4.9. It is skipped. Cilium v1.9 doesn't support Kubernetes 1.20 so we have to skip the test.

**Comment by Skymirrh** [comment]
> > Indeed, as can be seen either from automated jobs or PR jobs all passing without issue. I've confirmed it on local runs as well.
> 
> That's very strange. 🤔

You made me doubt so I double checked, and actually you are right: the test does not pass on 1.20 either. The difference is that it ends up as a skipped test instead of a fail, because Cilium v1.9 is not supported on K8s 1.20:

```
---
K8sUpdates Tests upgrade and downgrade from a Cilium stable image to master
at /workspace/git/cilium/test/ginkgo-ext/scopes.go:514

[terminating containers are not deleted after timeout
Expected
    <*errors.errorString | 0xc0002162e0>: {
        s: "Pods are still not deleted after a timeout: 4m0s timeout expired",
    }
to be nil]
21:29:28 STEP: Running AfterEach for block EntireTestsuite
<Checks>

</Checks>

Something has failed on AfterEach, running AfterFailed functions
21:29:28 STEP: Running AfterAll block for EntireTestsuite K8sUpdates

S [SKIPPING] in Spec Setup (BeforeEach) [1554.565 seconds]
K8sUpdates
/workspace/git//cilium/test/ginkgo-ext/scopes.go:478
  Tests upgrade and downgrade from a Cilium stable image to master [BeforeEach]
  /workspace/git/cilium/test/ginkgo-ext/scopes.go:514

  Cilium "v1.9" is not supported in K8s "1.20". Skipping upgrade/downgrade tests.

  /workspace/git/cilium/test/k8sT/Updates.go:65
------------------------------

```

**Comment by christarazi** [comment]
Should be fixed by https://github.com/cilium/cilium/pull/15099.

**Comment by Skymirrh** [comment]
Fixed by merge of #15099, closing.
